### PR TITLE
#9825 - Handle non ASCII command line parameters such as path

### LINKED ISSUE #9825: EnergyPlus fail to run when IDF file path has special character
Issue overview
--------------
EnergyPlus fail to run calculation when I have accented character in the path of the IDF file. For example, I have an IDF file in the folder "D:\térmica Oficinas". When I run simulation with EPLaunch, I get the following error:

image

If I run simulation in Command Line, I get the following error message:
image

### Details
Some additional details for this issue (if relevant):
 - Platform: Windows 10
 - Version of EnergyPlus: EnergyPlus 22.2.0
 - Unmethours link or helpdesk ticket number: N/A

### Checklist
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [ ] Defect file added (list location of defect file here)
 - [x] Ticket added to Pivotal for defect (development team task)
 - [ ] Pull request created (the pull request will have additional tasks related to reviewing changes that fix this defect)

### PR DESCRIPTION
**Pull request by jmarrec**
Pull request overview
---------------------
 - Fixes #9825
 - Replaces ezOptionParser with CLI11. Benefits of CLI11 include
     - does not use raw pointers
     - Cross platform (Really about windows...) to handle wide chars via narrowing 
         - Handling wide chars is possible, these are the early commits: https://github.com/NREL/EnergyPlus/pull/10120/files/9b97c017cd606dcc6748dbd14911d936f901d6a7..b19ec8829856a6d36cfa9b66f6133e842ce26edc
         - But CLI11 has this built in, tested, and it's an actively maintained project
      -  Lots of customization possible, validators, clear syntax, less boilerplate needed to achieve the same thing.

### Pull Request Author
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [ ] Title of PR should be user-synopsis style (clearly understandable in a standalone changelog context)
 - [ ] Label the PR with at least one of: Defect, Refactoring, NewFeature, Performance, and/or DoNoPublish
 - [ ] Pull requests that impact EnergyPlus code must also include unit tests to cover enhancement or defect repair
 - [ ] Author should provide a "walkthrough" of relevant code changes using a GitHub code review comment process
 - [ ] If any diffs are expected, author must demonstrate they are justified using plots and descriptions
 - [ ] If changes fix a defect, the fix should be demonstrated in plots and descriptions
 - [ ] If any defect files are updated to a more recent version, upload new versions here or on DevSupport
 - [ ] If IDD requires transition, transition source, rules, ExpandObjects, and IDFs must be updated, and add IDDChange label
 - [ ] If structural output changes, add to output rules file and add OutputChange label
 - [ ] If adding/removing any LaTeX docs or figures, update that document's CMakeLists file dependencies

### Reviewer
This will not be exhaustively relevant to every PR.
 - [ ] Perform a Code Review on GitHub
 - [ ] If branch is behind develop, merge develop and build locally to check for side effects of the merge
 - [ ] If defect, verify by running develop branch and reproducing defect, then running PR and reproducing fix
 - [ ] If feature, test running new feature, try creative ways to break it
 - [ ] CI status: all green or justified
 - [ ] Check that performance is not impacted (CI Linux results include performance check)
 - [ ] Run Unit Test(s) locally
 - [ ] Check any new function arguments for performance impacts
 - [ ] Verify IDF naming conventions and styles, memos and notes and defaults
 - [ ] If new idf included, locally check the err file and other outputs

### CODE REVIEW

File: src/EnergyPlus/CMakeLists.txt
Comment (jmarrec) [inline src/EnergyPlus/CMakeLists.txt:1111]: New ctests exercising the CLI with some accented filepaths. One without a PythonPlugin, one with

File: src/EnergyPlus/CommandLineInterface.hh
Comment (jmarrec) [inline src/EnergyPlus/CommandLineInterface.hh:76]: new signature

File: src/EnergyPlus/CommandLineInterface.cc
Comment (jmarrec) [inline src/EnergyPlus/CommandLineInterface.cc:105]: Some of the benefits of switching to CLI11 are there

File: src/EnergyPlus/CommandLineInterface.cc
Comment (jmarrec) [inline src/EnergyPlus/CommandLineInterface.cc:146]: Here's an example option. It pretty clear what's going on. The default `std::string prefixOutName` is set to "eplus", then it's reassigned by CLI11 **if** the non required option is passed.

File: src/EnergyPlus/CommandLineInterface.cc
Comment (jmarrec) [inline src/EnergyPlus/CommandLineInterface.cc:148]: Here's a flag.

File: src/EnergyPlus/CommandLineInterface.cc
Comment (jmarrec) [inline src/EnergyPlus/CommandLineInterface.cc:198]: This uses a check to ensure the file exists. It'll fail right away if the passed argument isn't an existing file (or if not passed if the default "in.epw" doesn't exist in the current directory)

File: src/ConvertInputFormat/main.cpp
Comment (jmarrec) [inline src/ConvertInputFormat/main.cpp:459]: ConvertInputFormat also gets revamped

File: cmake/CompilerFlags.cmake
Comment (jmarrec) [inline cmake/CompilerFlags.cmake:27]: https://learn.microsoft.com/en-us/cpp/build/reference/utf-8-set-source-and-executable-character-sets-to-utf-8?view=msvc-170

File: tst/EnergyPlus/unit/CommandLineInterface.unit.cc
Comment (jmarrec) [inline tst/EnergyPlus/unit/CommandLineInterface.unit.cc:242]: I added gtest for CommandLineInterface::ProcessArgs before, and they still pass after (minus some slight adjustments)

File: src/EnergyPlus/PluginManager.hh
Comment (jmarrec) [inline src/EnergyPlus/PluginManager.hh:187]: Removed sanitizedPath. (addToPythonPath I just renamed the parameter `path` -> `includePath`... path is a class, so it's kinda confusing.)

File: src/EnergyPlus/PluginManager.hh
Comment (jmarrec) [inline src/EnergyPlus/PluginManager.hh:215]: New

File: src/EnergyPlus/PluginManager.cc
Comment (jmarrec) [inline src/EnergyPlus/PluginManager.cc:1158]: addToPython path will now use:
 * Forward slashes
 * Handle a fs::path on windows correctly, where the fs::path::value_type is a `wchar_t`

File: src/EnergyPlus/PluginManager.cc
Comment (jmarrec) [inline src/EnergyPlus/PluginManager.cc:1173]: Use the Python C API to change sys.path instead of calling `PyRun_SimpleString`

File: src/EnergyPlus/PluginManager.cc
Comment (jmarrec) [inline src/EnergyPlus/PluginManager.cc:1136]: For debugging purposes.

File: src/EnergyPlus/PluginManager.cc
Comment (jmarrec) [inline src/EnergyPlus/PluginManager.cc:440]: * Handle the case where programDir is also non-ascii
* Don't leak the Py_DecodeLocal one. 

https://docs.python.org/3/c-api/sys.html#c.Py_DecodeLocale

> Return a pointer to a newly allocated wide character string, use [PyMem_RawFree()](https://docs.python.org/3/c-api/memory.html#c.PyMem_RawFree) to free the memory. If size is not NULL, write the number of wide characters excluding the null character into *size

File: src/EnergyPlus/PluginManager.cc
Comment (jmarrec) [inline src/EnergyPlus/PluginManager.cc:462]: The rest of the addToPythonPath calls need no special preparation since everything is handled in there. Smooth sailing

File: src/EnergyPlus/PluginManager.cc
Comment (jmarrec) [inline src/EnergyPlus/PluginManager.cc:476]: That was done twice. Here and in the `else` ladder of the `if (searchPaths > 0)` (much) below

File: src/EnergyPlus/PluginManager.cc
Comment (jmarrec) [inline src/EnergyPlus/PluginManager.cc:764]: Last `type_traits` used, I swear.

I'm technically doing this twice, here and in addToPythonPath, but I don't really want to create a helper that returns a `PyObject*` that the user is responsible to decref, and using a std::unique_ptr with a custom deleter or a wrapper object with a destructor that does Py_DECREF felt way overkilled for 2 occurrences (that, and you would need to use object.get() if a unique_ptr, or object.pyobj_ or something if a wrapper)

File: src/EnergyPlus/PluginManager.cc
Comment (jmarrec) [inline src/EnergyPlus/PluginManager.cc:770]: Better error message if we can't load the module, we print the content of of the sys.path

File: cmake/CompilerFlags.cmake
Comment (Myoldmopar) [reply cmake/CompilerFlags.cmake:27]: Interesting!  OK.

File: src/ConvertInputFormat/main.cpp
Comment (Myoldmopar) [inline src/ConvertInputFormat/main.cpp:428]: Hmmmmm I don't really like the C style inline comments.  Harder to parse.  Is there a reason to keep this in here commented at all?

File: src/ConvertInputFormat/main.cpp
Comment (Myoldmopar) [reply src/ConvertInputFormat/main.cpp:459]: :+1: Unexpected at first, but does make sense as it has its own CLI.

File: src/EnergyPlus/CMakeLists.txt
Comment (Myoldmopar) [reply src/EnergyPlus/CMakeLists.txt:1111]: Very nice.

File: src/EnergyPlus/CommandLineInterface.cc
Comment (Myoldmopar) [inline src/EnergyPlus/CommandLineInterface.cc:165]: I can see what's going on here.  It's a bit funny looking, and I would have a tough time just rattling off this code from memory.  But it's the CLI...it's not going to change almost ever.  And if it does, you've given plenty of examples we can base changes off of.  This is good.

File: src/EnergyPlus/CommandLineInterface.cc
Comment (Myoldmopar) [inline src/EnergyPlus/CommandLineInterface.cc:273]: This is a fine "hidden" flag to help diagnose issues.

File: src/EnergyPlus/PluginManager.cc
Comment (Myoldmopar) [reply src/EnergyPlus/PluginManager.cc:440]: `if constexpr` :+1: :+1: 

OK, this looks good.

File: src/EnergyPlus/PluginManager.cc
Comment (Myoldmopar) [reply src/EnergyPlus/PluginManager.cc:476]: OK, I'll have to think about the behavior here. The order that we add things to the path will ultimately determine the possible locations that Python code is found.  Any chance this change could cause different behavior?

File: src/EnergyPlus/PluginManager.cc
Comment (Myoldmopar) [inline src/EnergyPlus/PluginManager.cc:515]: It really needed you to tell it that epin_path was not const?  Weird.

File: src/EnergyPlus/PluginManager.cc
Comment (Myoldmopar) [inline src/EnergyPlus/PluginManager.cc:549]: :+1: a little more helpful, thanks :)

File: src/EnergyPlus/PluginManager.cc
Comment (Myoldmopar) [reply src/EnergyPlus/PluginManager.cc:764]: Yeah that's a reasonable justification for doing it twice.

File: src/EnergyPlus/PluginManager.cc
Comment (Myoldmopar) [reply src/EnergyPlus/PluginManager.cc:770]: Yes, very helpful!

File: src/EnergyPlus/PluginManager.cc
Comment (Myoldmopar) [reply src/EnergyPlus/PluginManager.cc:1173]: Oh, that's nice!

File: tst/EnergyPlus/unit/CMakeLists.txt
Comment (Myoldmopar) [inline tst/EnergyPlus/unit/CMakeLists.txt:293]: So we now have an energyplusapi_tests binary?  This is totally fine if so, and will be included when I run `ctest` either way, just wanting to confirm.

**Review by Myoldmopar — APPROVED** [review]
Holy cow this is a nice set of changes.  I have a few (very) minor questions/comments, but overall I like this.  CLI11 looks very nice, and you have made many changes to accompany this transition, and tested it heavily.

I will mark it approved, but I think my one comment about plugin path order is worth discussing before we move this toward merge.  Let me know what you think

File: src/ConvertInputFormat/main.cpp
Comment (jmarrec) [reply src/ConvertInputFormat/main.cpp:428]: Was trying to make it clearer that this is done on purpose, because I later defer to CLI11 to find the args. I can remove

File: src/EnergyPlus/CommandLineInterface.cc
Comment (jmarrec) [reply src/EnergyPlus/CommandLineInterface.cc:165]: Is this specific to this option or generally speaking?

File: src/EnergyPlus/PluginManager.cc
Comment (jmarrec) [reply src/EnergyPlus/PluginManager.cc:515]: clang-tidy can't see past the Optional here: https://github.com/NREL/EnergyPlus/blob/b7a30cf169f81fec733cee276d77bb4b8152dc64/third_party/ObjexxFCL/src/ObjexxFCL/environment.cc#L50-L54.

File: tst/EnergyPlus/unit/CMakeLists.txt
Comment (jmarrec) [reply tst/EnergyPlus/unit/CMakeLists.txt:293]: Yes. I was considering renaming it "energyplus_cli_tests" or something, but then that means I can't use create_test_targets, so either I have to modify it + every point that calls it, or I have to duplicate the code in the macro. So I figured that was fine...

File: src/EnergyPlus/PluginManager.cc
Comment (jmarrec) [reply src/EnergyPlus/PluginManager.cc:476]: before you had this:

```
if searchPaths == 0 {
   addToPythonPath('.')
}
if (searchPaths > 0) {
  [...]
} else {
  addToPythonPath('.')
}
```

File: src/EnergyPlus/CommandLineInterface.cc
Comment (Myoldmopar) [reply src/EnergyPlus/CommandLineInterface.cc:165]: Generally speaking.  The chain of trailing method calls honestly looks like..... Javascript.  `.then(function() { .then(function() ....`  It's totally fine and I'm not saying anything is wrong.  Just noting.

**Comment by Myoldmopar** [comment]
Thanks for commenting back.  The only thing I need to look at it is the order of the search path addition.  I'll check that out and see if I can get fully comfortable and then get this merged.

File: src/EnergyPlus/PluginManager.cc
Comment (Myoldmopar) [reply src/EnergyPlus/PluginManager.cc:476]: Ahhhhh I see...and the block below also got the `epin` environment variable tweak, making it better anyway.  OK, just had to see for myself.  I think this is good to go, thanks @jmarrec !

**Comment by Myoldmopar** [comment]
Alright, this looks good to go, if anyone has comments, speak quick or this is merging.

**Comment by Myoldmopar** [comment]
Thanks @jmarrec, merging.

**Comment by Myoldmopar** [comment]
OK, so reverting this temporarily.  It appeared all clean, but once it was merged, there were issues.  There are build warnings and also some failing tests.  I'll put links here to find them easily:
- https://myoldmopar.github.io/EnergyPlusBuildResults/EnergyPlus-7944eddb825c0e9dcef3bde63e45ec8288227321-Win64-Windows-10-VisualStudio-16.html#build
- https://myoldmopar.github.io/EnergyPlusBuildResults/EnergyPlus-7944eddb825c0e9dcef3bde63e45ec8288227321-x86_64-Linux-Ubuntu-22.04-gcc-11.3.html#build
- https://myoldmopar.github.io/EnergyPlusBuildResults/EnergyPlus-7944eddb825c0e9dcef3bde63e45ec8288227321-x86_64-Linux-Ubuntu-22.04-gcc-11.3-IntegrationCoverage-Debug.html#build
- https://myoldmopar.github.io/EnergyPlusBuildResults/EnergyPlus-7944eddb825c0e9dcef3bde63e45ec8288227321-x86_64-Linux-Ubuntu-22.04-gcc-11.3-UnitTestsCoverage-Debug.html#build
- https://myoldmopar.github.io/EnergyPlusBuildResults/EnergyPlus-7944eddb825c0e9dcef3bde63e45ec8288227321-x86_64-MacOS-10.17-clang-13.0.0.html#build

I'll actually try to tackle these myself if I can soon.  Regardless, this can be reviewed/merged again once it is all resolved.

**Comment by jmarrec** [comment]
@Myoldmopar `94810-1956.stat` exists, but `94810-1956.epw` does not. Same for `94810-1953-1957.stat`

I'm not sure what's the expected behavior here?

IMHO, we should throw if the EPW passed can't be found. And this becomes a Cmake issue here...

To be clear, "-D -w NONEXISTING.epw" is passed. And the current CLI doesn't care that the epw doesn't exist, because -D is used.

**Comment by Myoldmopar** [comment]
Oh interesting.  This is a good question.  I tend to agree -- if a user explicitly passes an EPW by command line, and it doesn't exist, we should probably stop and let them figure out what went wrong.  Even if the EPW was never going to be used.  

I'm open to discussion.  But what does it take to fix up our Cmake so that it works when enforced in this way?

**Comment by jmarrec** [comment]
I dug a bit deeper. 

It's not self evident, but actually passing `-D -w <epw>` even if the epw doesn't exist will currently do something different that passing nothing: **the stat file is parsed anyways**, and the location information is reported to the HTML / CSV tabular output in the "Weather Statistics File" section. Potentially this could also affect the SiteWaterMainsTemperature object too.

If you pass a non-existing EPW file and you do not pass the `-D` flag, we already throw a failure later and exit.

**Comment by jmarrec** [comment]
I will implement the same logic. Even though I think it's strange you'd have a stat file and not an epw file next to it, but whatever.

**Pull request merged**
